DM-24702 Write PiffPsf and PiffPsfDeterminer

jmeyers314 opened the pull request.

MorganSchmitz commented on tests/test_psf.py:
```diff
@@ -0,0 +1,278 @@
+
+        rand = afwMath.Random()  # make these tests repeatable by setting seed
+
+        addNoise = True
+
+        if addNoise:
```
I don't understand the point of those lines; can't we just get rid of the if statement altogether?

(Again, I know this is from meas_extensions_psfex!)

jmeyers314 replied:
Haha.  Yes, let's get rid of them.  Cargo culting at its best... :)

MorganSchmitz commented on python/lsst/meas/extensions/piff/piffPsf.py:
```diff
@@ -0,0 +1,99 @@
+    # Internal private methods
+
+    def _doImage(self, position, center):
+        # Follow Piff conventions for center.
+        # None => draw as if star at position
+        # True => draw in center of image
```
I kind of wish we could have this documented somewhere user-facing, though I see why it's hard since a user would only ever call the generic `computeImage`

jmeyers314 replied:
Yeah, I think the Psf documentation for `computeImage` vs `computeKernelImage` is probably sufficient for this one.

MorganSchmitz commented on python/lsst/meas/extensions/piff/piffPsfDeterminer.py:
```diff
@@ -0,0 +1,275 @@
+        self.kernelSizeMax = 35
+
+
+def computeWeight(maskedImage, maxSNR):
```
Is the math behind this (and ,maybe more importantly, `applyMaxSNR`) documented anywhere already?

jmeyers314 replied:
It's documented within Piff, but probably worth copying over here.  Will do.

MorganSchmitz reviewed (commented):
Looks great overall. I actually learned a lot going through this (like `afwImage.PARENT`), so thanks!

I do have a few questions/comments. A lot are on code that is pretty much the same as that of `meas_extensions_psfex`, and then there's the odd Piff questions in there - sorry!

My only serious concern is with the kernelSize, that I don't understand how we can choose to be different from the default value at present.

One last small thing: I see there's no `tearDown` method in the test. Is that intentional?

Great job with all this Josh, it's exciting to see Piff finally making its way to the stack!

jmeyers314 commented on python/lsst/meas/extensions/piff/piffPsf.py:
```diff
@@ -0,0 +1,99 @@
+    # ImagePsf overrides
+
+    def __deepcopy__(self, meta=None):
+        return PiffPsf(self.dimensions.x, self.dimensions.y, self.PiffResult)
```
That's an interesting point.  Stack `Psf` objects are supposed to be immutable, with a c++ `clone` method (which redirects to `__deepcopy__` in python) only available so that multiple caches can be used.  Of course you're right that if a user mutated `.piffResult` then the above wouldn't work for `__deepcopy__`.  We could make `._piffResult` "private", but I think even better would be to wrap `.piffResult` with a readonly property.

(Also, noticed that I mis-capitalized this!  So definitely an error on this line!)

jmeyers314 commented on python/lsst/meas/extensions/piff/piffPsfDeterminer.py:
```diff
@@ -0,0 +1,275 @@
+        dtype=float,
+        default=4.0
+    )
+    outlierMaxRemove = pexConfig.Field(
```
Only the worst offenders are removed.  The thinking is that the influence of some really bad outliers may make some good points look like outliers when they really shouldn't be.  So you only remove a few at a time.

jmeyers314 commented on python/lsst/meas/extensions/piff/piffPsfDeterminer.py:
```diff
@@ -0,0 +1,275 @@
+        }
+
+        piffResult = piff.PSF.process(piffConfig)
+        wcs = {0: galsim.PixelScale(1.0)}
```
You're correct!  And done.

jmeyers314 commented on python/lsst/meas/extensions/piff/piffPsfDeterminer.py:
```diff
@@ -0,0 +1,275 @@
+        default=200.0
+    )
+
+    def setDefaults(self):
+        self.kernelSize = 21
+        self.kernelSizeMin = 11
+        self.kernelSizeMax = 35
```
They are, these `pexConfig` fields are just defined in `BasePsfDeterminerConfig` instead of here :)

MorganSchmitz replied:
Ah, duh!

jmeyers314 commented on python/lsst/meas/extensions/piff/piffPsfDeterminer.py:
```diff
@@ -0,0 +1,275 @@
+        """
+        emi = exposure.getMaskedImage()
+        ebbox = emi.getBBox(afwImage.PARENT)
+        psfCellSet = afwMath.SpatialCellSet(
+            ebbox, self.config.sizeCellX, self.config.sizeCellY
+        )
```
It confuses me too, and I included it really only because it's one of the return values defined by `BasePsfDeterminerTask`.  It might be worth asking on #dm-psf or another slack channel whether this is worth keeping.

jmeyers314 replied:
Actually, looks like the cellSet is used more extensively at least in `pcaPsfDeterminer`, so there's at least that use case.  It might be better to just ignore it here though, similar to it's mostly ignored for PSFEx.

jmeyers314 replied:
I just went through and excised the cellSet stuff completely since it wasn't being used.  A cellSet is still part of the determiner return API, of course, but that's just hard-coded to None now (and documented as such).  I _think_ that's clearer.

jmeyers314 commented:
> One last small thing: I see there's no tearDown method in the test. Is that intentional?

Is one needed?  I do see there's a `tearDown` in `test_PsfexPsf`, which I obviously stole most of the test case from, but I kinda think this just frees memory that will be freed anyway when the test finishes.  If you know that that's wrong though, please tell me!

MorganSchmitz commented:
Hah! I thought I'd read somewhere (the DevGuide) it was better to always have a `tearDown`, but now I can't find it again. And there's precedent for test suites without it in the stack, so nevermind!

jmeyers314 commented:
Thanks for the review @MorganSchmitz !  I think I've addressed all of your comments at this point (and see you've +1'd most of them!).  Let me know if you have any further comments, especially about excising the psfCellSet stuff completely.

MorganSchmitz commented:
Yep, everything looks great now! Okay to merge whenever you're ready.
